include/vcml/models/dma/pl330.h

```diff
@@ -0,0 +1,279 @@
+            STOPPED = 0x0,
+            EXECUTING = 0x1,
+            CACHE_MISS = 0x2,
+            UPDATING_PC = 0x3,
+            WAITING_FOR_EVENT = 0x4,
+            AT_BARRIER = 0x5,
+            WAITING_FOR_PERIPHERAL = 0x7,
+            KILLING = 0x8,
+            COMPLETING = 0x9,
+            FAULTING_COMPLETING = 0xE,
+            FAULTING = 0xF,
+        };
+
+        enum fault : u8 {
+            UNDEF_INSTR = 0x0,
+            OPERAND_INVALID = 0x1,
+            CH_EVNT_ERR = 0x5,
+            CH_PERIPH_ERR = 0x6,
+            CH_RDWR_ERR = 0x7,
+            MFIFO_ERR = 0xC,
+            ST_DATA_UNAVAILABLE = 0xD,
+            INSTR_FETCH_ERR = 0x10,
+            DATA_WRITE_ERR = 0x11,
+            DATA_READ_ERR = 0x12,
+            DBG_INSTR = 0x1E,
+            LOCKUP_ERR = 0x1F,
+        };
+
+        reg<u32> ftr; // channel fault type register
+        reg<u32> csr; // channel status register
+        reg<u32> cpc; // channel pc register
+        reg<u32> sar; // source address register
+        reg<u32> dar; // destination address register
+        reg<u32> ccr; // channel control register
+        reg<u32> lc0; // loop counter 0 register
+        reg<u32> lc1; // loop counter 1 register
+
+        u32 tag; // aka channel id
+        bool stall;
+        u32 request_flag;
+        u32 watchdog_timer;
+
+        bool is_state(u8 state) const { return (get_state() == state); }
+        u32 get_state() const { return csr & 0x7; }
+        void set_state(u32 new_state) { csr = (csr & ~0x7) | new_state; }
+
+        virtual ~channel() = default;
+        VCML_KIND(dma::pl330::channel);
+        channel(const sc_module_name& nm, u32 tag);
+    };
+
+    class manager : public module
+    {
+    public:
+        enum state : u8 {
+            STOPPED = 0x0,
+            EXECUTING = 0x1,
+            CACHE_MISS = 0x2,
+            UPDATING_PC = 0x3,
+            WAITING_FOR_EVENT = 0x4,
+            FAULTING = 0xF,
+        };
+
+        enum fault : u8 {
+            UNDEF_INSTR = 0x0,
+            OPERAND_INVALID = 0x1,
+            DMAGO_ERR = 0x4,
+            EVNT_ERR = 0x5,
+            INSTR_FETCH_ERR = 0x10,
+            DBG_INSTR = 0x1e,
+        };
+
+        reg<u32> dsr;  // DMA Manager Status Register
+        reg<u32> dpc;  // DMA Program Counter Register
+        reg<u32> fsrd; // Fault Status DMA Manager Register
+        reg<u32> ftrd; // Fault Type DMA Manager Register
+
+        bool stall;
+        u32 watchdog_timer;
+
+        bool is_state(u8 state) const { return (get_state() == state); }
+        u32 get_state() const { return dsr & 0x7; }
+        void set_state(u32 new_state) { dsr = (dsr & ~0x7) | new_state; }
+
+        virtual ~manager() = default;
+        VCML_KIND(dma::pl330::manager);
+        manager(const sc_module_name& nm);
+    };
+
+    sc_event m_dma;
+
+    property<bool> enable_periph;
+    property<u32> num_channels;
+    property<u32> num_irq;
+    property<u32> num_periph;
```

Comment:
same as above, i think this could be determined from the number of ports in `periph_irq` that have been connected.

Reply:
this is how many are allowed to connect. 
Maybe this is something we don't/shouldn't model accurately? 
I wouldn't mind getting rid of this in favor of always using the max (32).
would that be preferred?